modules/ROOT/edits/dataweave-type-system.adoc

```diff
@@ -0,0 +1,358 @@
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+{
+    firstName: String,
+    lastName*: String,
+    age: Number
+}
+----
+
+For conditional fields, use `?`. For example, the following type allows the age field to be present or not:
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+{
+    firstName: String,
+    lastName*: String,
+    age?: Number
+}
+----
+
+[[close_open_objects]]
+==== Close and Open Objects
+
+A closed object means that the type only accepts an object value if there are no additional fields in the object,
+just those key-value pairs specified by the type.
+
+An open object only put constraints on the fields that are declared in the type, and if they are all verified, then the type accepts the value. The type accepts objects that have other fields apart of the ones mentioned in the declaration. All object types are open unless specified.
+
+The following example works even if `age` is not mentioned in the type declaration because the type of the variable `user` is an open object:
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+var user: {firstName: String, lastName: String} = {firstName: "John", lastName: "Smith", age: 34}
+----
+
+The syntax to specify a closed object is `{| |}`. The following example makes the `User` type support only the keys `firstName` and `lastName`. The script throws an exception because the `age` field is not accepted.
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+var user: {|firstName: String, lastName: String|} = {name: "John", lastName: "Smith", age: 34}
+----
+
+
+=== Function Type
+
+DataWeave is a functional language, and functions are first class citizens, which means that functions are values,
+and values have an associated type.
+
+The syntax to define a function type is:
+
+----
+(paramType: ParamType,...) -> ReturnType
+----
+
+For example, if you want to define in a constraint a `Function` type that has a parameter of type `String`, a second parameter of type `Number`, and a `Boolean` return value, the correct syntax is:
+
+----
+(paramA: String, paramB: Number) -> Boolean
+----
+
+For example, you can define a function that receives another function as an argument:
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+fun applyIDsChange(ids: Array<Number>, changeTo: (Number) -> Number): Array<Number> = ???
+----
+
+If you call the function `applyIDsChange` with a function that does not match the `changeTo` constraint `(Number) -> Number`, DataWeave throws a type-checking error.
+For example, the following call works because `abs` is a function that takes a value of the `Number` type and returns a `Number`, which matches the `changeTo` parameter.
+
+----
+applyIDsChange([1,-6, 3, -8], abs)
+----
+
+But this call fails because `sum` is a function that takes an Array and returns a Number, so it does not match our parameter constraint.
```

Comment:
```suggestion
But the following call fails because `sum` is a function that takes an Array and returns a Number, so it does not match the parameter constraint:
```

Reply:
`Array` 
`Number`
backticks around the types has become more important